tools/redcap2linst.php

```diff
@@ -0,0 +1,166 @@
+    ]
+);
+
+$inputFile = $options['f'] ?? $options['file'] ?? '';
+$outputDir = $options['output-dir'] ?? '';
+
+if (empty($inputFile) || empty($outputDir)) {
+    showHelp();
+    exit(1);
+}
+if (!is_dir($outputDir)) {
+    fprintf(STDERR, "Output directory $outputDir doesn't exist.\n");
+    exit(1);
+}
+if (!is_writeable($outputDir)) {
+    fprintf(STDERR, "Output directory $outputDir is not writeable.\n");
+    exit(1);
+}
+
+$fp = fopen($inputFile, "r");
+if ($fp === false) {
+    fprintf(STDERR, "Could not open file $inputFile\n");
+    exit(1);
+}
+
+$headers     = fgetcsv($fp);
+$instruments = [];
+while ($row = fgetcsv($fp)) {
+    $inst = $row[1];
+    if (!isset($instruments[$inst])) {
+        $instruments[$inst] = [];
+    }
+    $linstFormat = toLINST($row[3], $row[0], $row[4], $row[5]);
+    if (!empty($linstFormat)) {
+        $instruments[$inst][] = $linstFormat;
+    }
+}
+fclose($fp);
+
+outputFiles($outputDir, $instruments);
+
+/**
+ * Take a single line from the redcap dictionary and returns the
+ * closest LINST equivalent.
+ *
+ * @param string $redcaptype      The type from REDCap
+ * @param string $redcapfieldname The fieldname from REDCap
+ * @param string $redcaplabel     The label from REDCap
+ * @param string $redcapChoices   The choices columnfrom REDCap
+ *
+ * @return string
+ */
+function toLINST(
+    string $redcaptype,
+    string $redcapfieldname,
+    string $redcaplabel,
+    string $redcapChoices,
+) : string {
+    $label = str_replace($redcaplabel, "\n", "<br /><br />");
+    switch ($redcaptype) {
+    case 'text':
+        // text maps directly to LORIS
+        return "text{@}$redcapfieldname{@}$label";
+    case 'descriptive':
+        // descriptive maps to label with no field.
+        return "static{@}{@}$label";
+    case 'radio':
+    case 'dropdown':
+        // Radio or dropdown maps to a select and the options are in the
+        // same format in the dictionary.
+        return "select{@}$redcapfieldname{@}$label{@}"
+            . optionsToLINST($redcapChoices);
+    case 'checkbox':
+        // checkboxes are the same format as radios but allow multiple options,
+        // so map to a selectmultiple instead of a select
+        return "selectmultiple{@}$redcapfieldname{@}$label{@}"
+            . optionsToLINST($redcapChoices);
+    case 'yesno':
+        // Map yes/no fields to dropdowns with yes and no options.
+        return "select{@}$redcapfieldname{@}$label{@}"
+            . "NULL=>''{-}'yes'=>'Yes'{-}'no'=>'No'";
+    case 'calc':
+        // Calc maps to a score field. We create the DB field but don't do the score.
+        return "static{@}$redcapfieldname{@}$label";
+    case 'sql':
+        // The "SQL" data type seems to just be for presentation? I hope?
+        return "";
+    case 'notes':
+        // REDCap calls textareas notes
+        return "textarea{@}$redcapfieldname{@}$label";
+    default:
+        throw new \LorisException("Unhandled REDCap type $redcaptype");
```

Comment:
```
PHP Fatal error:  Uncaught LorisException: Unhandled REDCap type file in /var/www/loris/tools/redcap2linst.php:103
Stack trace:
#0 /var/www/loris/tools/redcap2linst.php(44): toLINST()
#1 {main}
  thrown in /var/www/loris/tools/redcap2linst.php on line 103
```

could we print out the type and fieldname here?

Reply:
just noticed the type is already printed out ... i have a 'file' type in my dictionary

Reply:
also have 'slider' type

Reply:
Are sliders always numeric in redcap? Do we need to parse the dictionary options column or anything? Is the min/max customizable? Maybe you can send me a line from a CSV that's an example of one.

Not sure about file. @ridz1208 do you know if we can do a file upload type in LINST? I don't see one in the code but I thought we had one..

Reply:
i'm not sure whether sliders are all numeric! i would assume so but only one of my slider elements explicitly say "numeric" under the "Text Validation Type OR Show Slider Number" column.

```
qod_annoying | questionnaire_for_olfactory_disorders |   | slider | 30. Please use the scale below to rate how annoying the changes in your sense of smell are to you | Not annoying at all \| Extremely annoying |
```

Reply:
the min/max isn't specified in the dictionary, i just know its 0 to 100 cz copn user told me so i would assume its fixed